functions

```diff
@@ -336,58 +336,47 @@ add_checked_modules() {
 }
 
 add_firmware() {
-    # add a firmware file to the image.
-    #   $1: firmware path fragment
+    local fw p
 
-    local fw fwpath r=1
+    [[ $KERNELVERSION == none ]] && return
 
     for fw; do
-        for fwpath in "${_d_firmware[@]}"; do
-            if [[ -f $fwpath/$fw ]]; then
-                add_file "$fwpath/$fw" "$fwpath/$fw" 644 && r=0
-                break
-            fi
+        for p in "$_d_firmware"{/updates,}; do
+            [[ -f "$p/$fw" ]] || continue
+            add_file "$p/$fw" "/usr/lib/firmware/$fw" 0644
+            continue 2
         done
+        warning 'missing firmware: %s' "$fw"
     done
-
-    return $r
 }
 
-add_module() {
-    # Add a kernel module to the initcpio image. Dependencies will be
-    # discovered and added.
-    #   $1: module name
-
-    local target= module= softdeps= deps= field= value= firmware=()
-    local ign_errors=0 found=0
+_add_module() {
+    local m module softdeps deps field value
 
     [[ $KERNELVERSION == none ]] && return 0
 
-    if [[ $1 = *\? ]]; then
-        ign_errors=1
-        set -- "${1%?}"
-    fi
-
-    target=${1%.ko*} target=${target//-/_}
-
-    # skip expensive stuff if this module has already been added
-    (( _addedmodules["$target"] > 0 )) && return
+    _add_modules=()
 
     while IFS=':= ' read -r -d '' field value; do
         case "$field" in
+            name)
+                quiet "module:   %s" "$value"
+                ;;
             filename)
-                found=1
+                quiet "filename: %s" "$value"
                 module=${value##*/} module=${module%.ko*}
-                quiet "adding module: %s" "$module"
                 _modpaths["$value"]=1
                 _addedmodules["${module//-/_}"]=1
                 ;;
+            alias)
+                _addedmodules["${value//-/_}"]=3
```

Comment:
The only place where this is used outside of module resolution is parse_config() where it won't add builtins, and now aliases to MODULES. Currently this is not used anywhere to detect aliases so it could be changed to 2.